aya/src/symbols.rs

```diff
@@ -0,0 +1,688 @@
+    pub fn unresolved_kernel(address: u64) -> Self {
+        Self {
+            virtual_address: address,
+            object_address: Some(address),
+            function_name: None,
+            object_path: None,
+            process_id: None,
+        }
+    }
+
+    /// Creates a `SymbolInfo` instance for a user-space function which hasn't been resolved yet
+    pub fn unresolved_user(process_id: u32, address: u64) -> Self {
+        Self {
+            virtual_address: address,
+            object_address: None,
+            function_name: None,
+            object_path: None,
+            process_id: Some(process_id),
+        }
+    }
+
+    /// Whether the symbol refers to a kernel-space function
+    pub fn is_kernel(&self) -> bool {
+        self.process_id.is_none()
+    }
+
+    /// Whether the symbol refers to a user-space function
+    pub fn is_user(&self) -> bool {
+        self.process_id.is_some()
+    }
+
+    /// Returns the ID of the process for which the symbol was recorded
+    pub fn process_id(&self) -> Option<u32> {
+        self.process_id
+    }
+
+    /// Virtual memory address of a certain function
+    pub fn virtual_address(&self) -> u64 {
+        self.virtual_address
+    }
+
+    /// Address of a function inside an object. `None` if it was not resolved
+    pub fn object_address(&self) -> Option<u64> {
+        self.object_address
+    }
+
+    /// Path to the object which defines the symbol
+    pub fn object_path(&self) -> Option<&Path> {
+        self.object_path.as_deref()
+    }
+
+    /// The resolved function name
+    pub fn function_name(&self) -> Option<&str> {
+        self.function_name.as_deref()
+    }
+}
+
+/// Resolves a symbol based on it's address.
+///
+/// See [`DefaultResolver`] for exemplaric usage.
+pub trait SymbolResolver {
+    /// Resolves a symbol based on it's address
+    fn resolve(&self, symbol: &mut SymbolInfo);
+}
+
+/// A resolver for kernel symbols
+pub struct KernelSymbolResolver {
+    symbols: BTreeMap<u64, String>,
+}
+
+impl KernelSymbolResolver {
+    /// Creates a new `KernelSymbolResolver`
+    ///
+    /// This will load all kernel symbols from `/proc/kallsyms`.
+    pub fn new() -> Result<Self, CreateResolverError> {
+        let symbols = kernel_symbols()
+            .map_err(|e| CreateResolverError::KernelSymbolResolverBuildError { error: e })?;
+        Ok(Self { symbols })
+    }
+}
+
+impl SymbolResolver for KernelSymbolResolver {
+    fn resolve(&self, symbol: &mut SymbolInfo) {
+        if symbol.process_id().is_some() || symbol.function_name.is_some() {
+            return;
+        }
+
+        symbol.function_name = self
+            .symbols
+            .range(..=symbol.virtual_address())
+            .next_back()
+            .map(|(_, s)| s.clone());
+    }
+}
+
+/// A SymbolResolver which uses the `addr2line` Rust library
+pub struct Addr2LineResolver {
```

Comment:
I think this is probably not the right level of abstraction. I'm not sure we should provide an API that can resolve from multiple pids and multiple object files. I think we should provide primitives and then let people could compose them as they wish. Eg, we should definitely provide the primitives to resolve from kernel, object files and pids - but should we provide a full blown API to implement a system wide profiler that can profile kernel and separate processes? I'm inclined to say no.

Generally speaking, I see aya as a high level rust wrapper for the primitives offered by the ebpf platform. Higher level abstractions that can be built using core primitives should ideally go in side crates. I do think that it would be really cool to provide a complete profiler API, but it should probably go in a `aya-profiler` crate or something.

Reply:
> . Eg, we should definitely provide the primitives to resolve from kernel, object files and pids

I agree - that's why all basic resolvers are also individually available

> but should we provide a full blown API to implement a system wide profiler that can profile kernel and separate processes? I'm inclined to say no.

I think yes - just because it requires a significant amount of domain knowledge to get a full stack trace for whatever values the BPF code returns otherwise. I certainly spent multiple hours on topics that are 100% unrelated to both eBPF and the application I wanted to look into  before I got any useful results. I think eBPF application developers will often want to get a readable stack trace, but only the minority might be interested in how the innards of that work.

Also consider that those functions are not only interesting for profiler authors that might have a bit of a better background around call stacks - but also for all others. E.g. I might want to write a program for debugging purposes which logs the time spent on particular syscalls - and we then want to figure out which codepath actually triggered high latency. A simple resolver with a default constructor that allows to act on the address/pid values that the BPF code submits can do that job - and exactly that is available via `DefaultResolver`. 

I assume other tools like `bpftrace` have the necessary code included for  a similar reason.

It's however a fair question to ask which "intermediate" layer tools should be provided, and e.g. whether the `struct Addr2LineResolver` with 2 layers of caching here really should be public - instead of being purely an implementation detail of  `DefaultResolver` (or DefaultUserspaceResolver` if you want to separate those).

If only building blocks and the high level resolver would be public, it would be:
- `KernelResolver`
- `ProcMemMap`
- `Addr2LineObjectContext`  (potentially to be renamed)
- `DefaultResolver` (potentially to be renamed) 

And since most of those don't have common input/result types anymore, it might not be worthwhile to have the `Resolver` trait anymore.